source/code/plugins/VMInsightsDataCollector.rb

```diff
@@ -0,0 +1,393 @@
+
+        def get_disk_stats(dev)
+            raise @baseline_exception if @baseline_exception
+            @saved_disk_data.get_disk_stats(dev)
+        end
+
+    private
+
+        class DataWithWrappingCounter
+            @@counter_modulus = 0   # default to cause exception
+
+            def self.set_32_bit(is32bit)
+                @@counter_modulus = (2 ** (is32bit ? 32 : 64))
+            end
+
+        protected
+
+            def sub_with_wrap(a, b)
+                (@@counter_modulus + a - b) % @@counter_modulus
+            end
+
+        end
+
+        class DiskInventory
+            def initialize(root)
+                @root = root
+                @sector_sizes = Hash.new() { |h, k| h[k] = get_sector_size(k) }
+                @saved_disk_data = { }
+            end
+
+            def baseline
+                @sector_sizes.replace(get_sector_sizes)
+                @saved_disk_data = { }
+                @sector_sizes.each_pair { |d, s| @saved_disk_data[d] = get_disk_data(d, s) }
+            end
+
+            def get_disk_stats(dev)
+                current = get_disk_data dev, get_sector_size(dev)
+                raise IDataCollector::Unavailable, "no data for #{dev}" if current.nil?
+                previous = @saved_disk_data[dev]
+                @saved_disk_data[dev] = current
+                raise IDataCollector::Unavailable, "no previous data for #{dev}" if previous.nil?
+                current - previous
+            end
+
+        private
+
+            def get_sector_size(dev)
+                raise ArgumentError, "dev is nil" if dev.nil?
+                data = get_sector_sizes(dev)
+                data[dev]
+            end
+
+            def get_sector_sizes(*devices)
+                cmd = [ File.join(@root, "bin", "lsblk"), "-sd", "-oNAME,LOG-SEC" ]
+                result = { }
+                begin
+                    IO.popen(cmd, { :in => :close }) { |io|
+                        io.gets # skips the header
+                        while (line = io.gets)
+                            s = line.split(" ")
+                            next if s.length < 2
+                            if (devices.empty? || devices.include?(s[0]))
+                                result[s[0]] = s[1].to_i
+                            end
+                        end
+                    }
+                rescue Errno::ENOENT => ex
+                    # telemetry
+                end
+                result
+            end
+
+            def get_disk_data(dev, sector_size)
+                path = File.join(@root, "sys", "class", "block", dev, "stat")
+                File.open(path, "rb") { |f|
+                    line = f.gets
+                    raise Unavailable, "#{path}: is empty" if line.nil?
+                    data = line.split(" ")
+                    RawDiskData.new(
+                                    dev,
+                                    Time.now,
+                                    data[1 - 1].to_i,
```

Comment:
What is the idea of expressing the index like this, rather than 0, 2, 4, and 6?

Reply:
The documentation I was referencing, which I failed to cite (bad on me) used one based column numbering.
Will change before check-in.